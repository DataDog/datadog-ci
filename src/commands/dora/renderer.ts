import chalk from 'chalk'

const ICONS = {
  FAILED: '❌',
  SUCCESS: '✅',
  WARNING: '⚠️',
}

export const renderFailedRequest = (service: string, errorMessage: string) => {
  return chalk.red(`${ICONS.FAILED} Failed to send DORA deployment event for service: ${service}: ${errorMessage}\n`)
}

export const renderRetriedRequest = (service: string, errorMessage: string, attempt: number) => {
  return chalk.yellow(
    `[attempt ${attempt}] Retrying to send DORA deployment event for service: ${service}: ${errorMessage}\n`
  )
}

export const renderSuccessfulRequest = (service: string) => {
  return chalk.green(`${ICONS.SUCCESS} Successfuly sent DORA deployment event for service: ${service}`)
}

export const renderDryRun = (service: string): string => `[DRYRUN] ${renderRequest(service)}`

export const renderRequest = (service: string): string => `Sending DORA deployment event for service: ${service}`
