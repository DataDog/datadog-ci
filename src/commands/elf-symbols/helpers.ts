import {getBaseSourcemapIntakeUrl} from '../../helpers/base-intake-url'
import {RequestBuilder} from '../../helpers/interfaces'
import {MultipartPayload, upload, UploadOptions} from '../../helpers/upload'
import {getRequestBuilder} from '../../helpers/utils'

export const getElfRequestBuilder = (apiKey: string, cliVersion: string, site: string) =>
  getRequestBuilder({
    apiKey,
    baseUrl: getBaseSourcemapIntakeUrl(site),
    headers: new Map([
      ['DD-EVP-ORIGIN', 'datadog-ci_elf-symbols'],
      ['DD-EVP-ORIGIN-VERSION', cliVersion],
    ]),
    overrideUrl: 'api/v2/srcmap',
  })

// This function exists partially just to make mocking network calls easier.
export const uploadMultipartHelper = async (
  requestBuilder: RequestBuilder,
  payload: MultipartPayload,
  opts: UploadOptions
) => upload(requestBuilder)(payload, opts)
