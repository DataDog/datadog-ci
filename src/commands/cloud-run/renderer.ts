import * as helpersRenderer from '../../helpers/renderer'

const AUTHENTICATION_INSTRUCTIONS = [
  '\n' + helpersRenderer.renderError('Unable to authenticate with GCP.'),
  'To authenticate with GCP, please follow these steps:',
  "1. If you haven't already, install the Google Cloud SDK: https://cloud.google.com/sdk/docs/install",
  '2. Run "gcloud auth application-default login" and follow the prompts in your browser to log in.',
  '3. After logging in, run the `datadog-ci cloud-run flare` command again.\n',
]
/**
 * @returns instructions on how to authenticate with GCP.
 */
export const renderAuthenticationInstructions = () => {
  return AUTHENTICATION_INSTRUCTIONS.join('\n')
}
