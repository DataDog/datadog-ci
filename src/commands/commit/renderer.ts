import chalk from 'chalk'

import {ICONS} from '../../helpers/formatting'

export const renderGitError = (errorMessage: string) =>
  chalk.red(`${ICONS.FAILED} An error occured while invoking git: ${errorMessage}
Make sure the command is running within your git repository.\n`)

export const renderConfigurationError = (error: Error) => chalk.red(`${ICONS.FAILED} Configuration error: ${error}.\n`)

export const renderFailedUpload = (errorMessage: string) =>
  chalk.red(`${ICONS.FAILED} Failed upload: ${errorMessage}\n`)

export const renderRetriedUpload = (errorMessage: string, attempt: number) =>
  chalk.yellow(`[attempt ${attempt}] Retrying upload: ${errorMessage}\n`)

export const renderUpload = 'Uploading\n'

export const renderSuccessfulCommand = (duration: number, dryRun: boolean) => {
  if (dryRun) {
    return chalk.green(`${ICONS.SUCCESS} [DRYRUN] Handled in ${duration} seconds.\n`)
  } else {
    return chalk.green(`${ICONS.SUCCESS} Uploaded in ${duration} seconds.\n`)
  }
}

export const renderCommandInfo = (dryRun: boolean) => {
  if (dryRun) {
    return chalk.yellow(`${ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD\n`)
  }

  return chalk.green('Starting upload. \n')
}
