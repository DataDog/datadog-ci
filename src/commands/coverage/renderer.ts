import chalk from 'chalk'
import upath from 'upath'

import {SpanTags} from '../../helpers/interfaces'

import {Payload} from './interfaces'
import {getCoverageDetailsUrl} from './utils'

const ICONS = {
  FAILED: '❌',
  SUCCESS: '✅',
  WARNING: '⚠️',
}

export const renderFileReadError = (filePath: string, errorMessage: string) => {
  const reportPath = `[${chalk.bold.dim(filePath)}]`

  return chalk.red(`${ICONS.FAILED} Error while reading report file ${reportPath}: ${errorMessage}\n`)
}

export const renderInvalidFile = (filePath: string, errorMessage: string) => {
  const reportPath = `[${chalk.bold.dim(filePath)}]`

  return chalk.red(`${ICONS.FAILED} Invalid coverage report file ${reportPath}: ${errorMessage}\n`)
}

export const renderFailedUpload = (payload: Payload, errorMessage: string) => {
  const payloadDetails = `${chalk.bold.dim(payload.paths)}`

  return chalk.red(`${ICONS.FAILED} Upload failed for ${payloadDetails}: ${errorMessage}\n`)
}

export const renderRetriedUpload = (payload: Payload, errorMessage: string, attempt: number) => {
  const payloadDetails = `${chalk.bold.dim(payload.paths)}`

  return chalk.yellow(`[attempt ${attempt}] Retrying coverage report upload ${payloadDetails}: ${errorMessage}\n`)
}

export const renderSuccessfulUpload = (dryRun: boolean, fileCount: number, duration: number) => {
  return chalk.green(`${dryRun ? '[DRYRUN] ' : ''}${ICONS.SUCCESS} Uploaded ${fileCount} files in ${duration} seconds.`)
}

export const renderSuccessfulUploadCommand = (spanTags: SpanTags) => {
  const coverageDetailsUrl = getCoverageDetailsUrl(spanTags)
  if (coverageDetailsUrl) {
    let fullStr = ''
    fullStr += chalk.green(
      '=================================================================================================\n'
    )
    fullStr += chalk.green(
      '* View detailed coverage report in Datadog (it can take a few minutes to become available)\n'
    )
    fullStr += chalk.green(`* ${coverageDetailsUrl}\n`)
    fullStr += chalk.green(
      '=================================================================================================\n'
    )

    return fullStr
  }

  return ''
}

export const renderDryRunUpload = (payload: Payload): string => `[DRYRUN] ${renderUpload(payload)}`

export const renderUpload = (payload: Payload): string => {
  if (payload.paths && payload.paths.length) {
    return `Uploading code coverage report file(s) in ${payload.paths}`
  } else {
    return 'No code coverage report paths, doing nothing'
  }
}

export const renderCommandInfo = (basePaths: string[], dryRun: boolean) => {
  let fullStr = ''
  if (dryRun) {
    fullStr += chalk.yellow(`${ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD COVERAGE REPORTS\n`)
  }
  fullStr += chalk.green(`${new Date().toLocaleString()} - Starting upload. \n`)
  if (!!basePaths.length) {
    if (basePaths.length === 1 && !!upath.extname(basePaths[0])) {
      fullStr += chalk.green(`Will upload code coverage report file ${basePaths[0]}`)
    } else {
      fullStr += chalk.green(`Will look for code coverage report files in ${basePaths.join(', ')}`)
    }
  }

  return fullStr
}

export const renderSuccessfulGitDBSync = (dryRun: boolean, elapsed: number) => {
  return chalk.green(`${dryRun ? '[DRYRUN] ' : ''}${ICONS.SUCCESS} Synced git metadata in ${elapsed} seconds.`)
}

export const renderFailedGitDBSync = (err: any) => {
  return chalk.red.bold(`${ICONS.FAILED} Could not sync git metadata: ${err}\n`)
}
