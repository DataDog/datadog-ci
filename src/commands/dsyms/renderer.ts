import path from 'path'

import chalk from 'chalk'

import {ICONS} from '../../helpers/formatting'
import {UploadStatus} from '../../helpers/upload'

import {ArchSlice, CompressedDsym, Dsym} from './interfaces'
import {pluralize} from './utils'

export const renderConfigurationError = (error: Error) => chalk.red(`${ICONS.FAILED} Configuration error: ${error}.\n`)

export const renderInvalidDsymWarning = (dSYMPath: string) =>
  chalk.yellow(`${ICONS.WARNING} Invalid dSYM file, will be skipped: ${dSYMPath}\n`)

export const renderDSYMSlimmingFailure = (dSYM: Dsym, slice: ArchSlice) =>
  chalk.yellow(`${ICONS.WARNING} Failed to export '${slice.arch}' arch (${slice.uuid}) from ${dSYM.bundlePath}\n`)

export const renderFailedUpload = (dSYM: CompressedDsym, errorMessage: string) => {
  const dSYMPathBold = `[${chalk.bold.dim(dSYM.dsym.bundlePath)}]`

  return chalk.red(`${ICONS.FAILED} Failed upload dSYM for ${dSYMPathBold}: ${errorMessage}\n`)
}

export const renderRetriedUpload = (dSYM: CompressedDsym, errorMessage: string, attempt: number) => {
  const dSYMPathBold = `[${chalk.bold.dim(dSYM.dsym.bundlePath)}]`

  return chalk.yellow(`[attempt ${attempt}] Retrying dSYM upload ${dSYMPathBold}: ${errorMessage}\n`)
}

export const renderSuccessfulCommand = (statuses: UploadStatus[], duration: number, dryRun: boolean) => {
  const results = new Map<UploadStatus, number>()
  statuses.forEach((status) => {
    if (!results.has(status)) {
      results.set(status, 0)
    }
    results.set(status, results.get(status)! + 1)
  })

  const output = ['', chalk.bold('Command summary:')]
  if (results.get(UploadStatus.Failure)) {
    output.push(chalk.red(`${ICONS.FAILED} Some dSYMS have not been uploaded correctly.`))
  } else if (results.get(UploadStatus.Skipped)) {
    output.push(chalk.yellow(`${ICONS.WARNING}  Some dSYMs have been skipped.`))
  } else if (results.get(UploadStatus.Success)) {
    if (dryRun) {
      output.push(
        chalk.green(
          `${ICONS.SUCCESS} [DRYRUN] Handled ${pluralize(
            results.get(UploadStatus.Success)!,
            'dSYM',
            'dSYMs'
          )} with success in ${duration} seconds.`
        )
      )
    } else {
      output.push(
        chalk.green(
          `${ICONS.SUCCESS} Uploaded ${pluralize(
            results.get(UploadStatus.Success)!,
            'dSYM',
            'dSYMs'
          )} in ${duration} seconds.`
        )
      )
    }
  } else {
    output.push(chalk.yellow(`${ICONS.WARNING} No dSYMs detected. Did you specify the correct directory?`))
  }

  if (results.get(UploadStatus.Failure) || results.get(UploadStatus.Skipped)) {
    output.push(`Details about the found ${pluralize(statuses.length, 'dSYM', 'dSYMs')}:`)
    if (results.get(UploadStatus.Success)) {
      output.push(`  * ${pluralize(results.get(UploadStatus.Success)!, 'dSYM', 'dSYMs')} successfully uploaded`)
    }
    if (results.get(UploadStatus.Skipped)) {
      output.push(
        chalk.yellow(`  * ${pluralize(results.get(UploadStatus.Skipped)!, 'dSYM was', 'dSYMs were')} skipped`)
      )
    }
    if (results.get(UploadStatus.Failure)) {
      output.push(chalk.red(`  * ${pluralize(results.get(UploadStatus.Failure)!, 'dSYM', 'dSYMs')} failed to upload`))
    }
  }

  return output.join('\n') + '\n'
}

export const renderCommandInfo = (basePath: string, poolLimit: number, dryRun: boolean) => {
  let fullStr = ''
  if (dryRun) {
    fullStr += chalk.yellow(`${ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD DSYMS\n`)
  }
  const startStr = chalk.green(`Starting upload with concurrency ${poolLimit}. \n`)
  fullStr += startStr
  const basePathStr = chalk.green(`Will look for dSYMs in ${basePath}\n`)
  fullStr += basePathStr

  fullStr += chalk.green(
    `Once dSYMs upload is successful files will be processed and ready to use within the next 5 minutes.\n`
  )

  return fullStr
}

export const renderCommandDetail = (intermediateDirectory: string, uploadDirectory: string) =>
  `Will use temporary intermediate directory: ${intermediateDirectory}\n` +
  `Will use temporary upload directory: ${uploadDirectory}\n`

export const renderUpload = (dSYM: CompressedDsym): string => {
  const archiveName = path.basename(dSYM.archivePath)
  const objectName = dSYM.dsym.slices.map((slice) => path.basename(slice.objectPath))[0]
  const archs = dSYM.dsym.slices.map((slice) => slice.arch).join()
  const uuids = dSYM.dsym.slices.map((slice) => slice.uuid).join()

  return `Uploading ${archiveName} (${objectName}, arch: ${archs}, UUID: ${uuids})\n`
}
