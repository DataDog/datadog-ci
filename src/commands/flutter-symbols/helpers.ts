import * as path from 'path'

import {getBaseSourcemapIntakeUrl} from '../../helpers/base-intake-url'
import {RequestBuilder} from '../../helpers/interfaces'
import {MultipartPayload, upload, UploadOptions} from '../../helpers/upload'
import {getRequestBuilder} from '../../helpers/utils'

export const getFlutterRequestBuilder = (apiKey: string, cliVersion: string, site: string) =>
  getRequestBuilder({
    apiKey,
    baseUrl: getBaseSourcemapIntakeUrl(site),
    headers: new Map([
      ['DD-EVP-ORIGIN', 'datadog-ci flutter-symbols'],
      ['DD-EVP-ORIGIN-VERSION', cliVersion],
    ]),
    overrideUrl: 'api/v2/srcmap',
  })

// This function exists partially just to make mocking networkc calls easier.
export const uploadMultipartHelper = async (
  requestBuilder: RequestBuilder,
  payload: MultipartPayload,
  opts: UploadOptions
) => upload(requestBuilder)(payload, opts)

export const getArchInfoFromFilename = (filename: string) => {
  const parsed = path.parse(filename)
  const basename = parsed.name
  const groups = /^.*\.(?<platform>.*)-(?<arch>.*)$/.exec(basename)?.groups
  if (!groups) {
    return undefined
  }

  const value = {
    arch: groups.arch,
    platform: groups.platform,
  }

  if (!value.platform || !value.arch) {
    return undefined
  }

  return value
}
