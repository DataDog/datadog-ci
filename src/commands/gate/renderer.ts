import chalk from 'chalk'

import {GIT_BRANCH, GIT_REPOSITORY_URL} from '../../helpers/tags'

import {EvaluationResponse, Payload, RuleEvaluation} from './interfaces'
import {getStatus, is5xxError, isBadRequestError, isTimeout, getBaseUrl} from './utils'

const ICONS = {
  FAILED: '❌',
  SUCCESS: '✅',
  WARNING: '⚠️',
  INFO: 'ℹ️',
}

export const renderEvaluationResponse = (evaluationResponse: EvaluationResponse) => {
  if (evaluationResponse.status.toLowerCase() === 'empty') {
    return renderEmptyEvaluation()
  }
  if (evaluationResponse.status.toLowerCase() === 'dry_run') {
    return renderDryRunEvaluation(evaluationResponse)
  }

  let fullStr = ''
  fullStr += chalk.green('Successfully evaluated all matching rules.\n')
  fullStr += `Overall result: ${renderStatus(evaluationResponse.status)}\n`
  fullStr += `Number of rules evaluated: ${chalk.bold(evaluationResponse.rule_evaluations.length)}\n`

  fullStr += '\n'
  fullStr += chalk.yellow('####### Rules evaluated #######\n')
  evaluationResponse.rule_evaluations.forEach((ruleEvaluation) => (fullStr += renderRuleEvaluation(ruleEvaluation)))

  return fullStr
}

export const renderEmptyEvaluation = (): string => {
  return chalk.yellow(
    `${ICONS.WARNING} No matching rules were found in Datadog. Use the '--fail-on-empty' option to fail the command in this situation.\n`
  )
}

export const renderStatus = (result: string): string => {
  switch (result.toLowerCase()) {
    case 'passed':
      return chalk.green(`Passed ${ICONS.SUCCESS} `)
    case 'failed':
      return chalk.red(`Failed ${ICONS.FAILED} `)
    case 'no_data':
      return chalk.yellow(`No Data ${ICONS.WARNING} `)
    case 'dry_run':
      return chalk.yellow(`Dry Run ${ICONS.INFO}`)
  }

  return result.toLowerCase()
}

export const renderRuleUrl = (ruleId: string): string => {
  return `${getBaseUrl()}ci/quality-gates/rule/${ruleId}`
}

export const renderRuleEvaluation = (ruleEvaluation: RuleEvaluation): string => {
  let fullStr = ''
  fullStr += `Rule Name: ${ruleEvaluation.rule_name}\n`
  fullStr += `Rule URL: ${renderRuleUrl(ruleEvaluation.rule_id)}\n`
  fullStr += `Status: ${renderStatus(ruleEvaluation.status)}\n`
  if (ruleEvaluation.status.toLowerCase() === 'failed') {
    fullStr += `${chalk.red.bold('Failure reason')}: ${ruleEvaluation.failure_reason}\n`
  }

  fullStr += `${chalk.yellow('Blocking')}: ${ruleEvaluation.is_blocking}\n`
  if (ruleEvaluation.details_url) {
    fullStr += `Details: ${encodeURI(ruleEvaluation.details_url)}\n`
  }

  fullStr += '\n'

  return fullStr
}

export const renderDryRunEvaluation = (evaluationResponse: EvaluationResponse): string => {
  let fullStr = ''
  fullStr += chalk.green('Successfully completed a dry run request\n')
  fullStr += `Overall result: ${renderStatus(evaluationResponse.status)}\n`
  fullStr += `Number of matching rules: ${chalk.bold(evaluationResponse.rule_evaluations.length)}\n`

  if (evaluationResponse.rule_evaluations.length > 0) {
    fullStr += '\n'
    fullStr += chalk.yellow('####### Matching rules #######\n')
    evaluationResponse.rule_evaluations.forEach((ruleEvaluation) => (fullStr += renderRuleEvaluation(ruleEvaluation)))
  }

  return fullStr
}

export const renderGateEvaluationInput = (evaluateRequest: Payload): string => {
  let fullStr = chalk.bold(`${ICONS.INFO} Evaluating rules matching the following information:\n`)
  fullStr += `Repository: ${evaluateRequest.spanTags[GIT_REPOSITORY_URL]}\n`
  fullStr += `Branch: ${evaluateRequest.spanTags[GIT_BRANCH]}\n`

  for (const [scopeKey, scopeValue] of Object.entries(evaluateRequest.userScope)) {
    const valueString = scopeValue.join(' OR ')
    fullStr += `${scopeKey}: ${valueString}\n`
  }

  fullStr += '\n'

  return fullStr
}

export const renderGateEvaluationError = (error: any, failIfUnavailable: boolean): string => {
  let errorStr = 'ERROR: Could not evaluate the rules.'

  if (error.message === 'wait') {
    errorStr += ` The command timed out.\n`
  }

  if (error.response) {
    errorStr += ` Status code: ${error.response.status}.\n`
  }

  if (isBadRequestError(error)) {
    const errorMessage = error.response.data.errors[0].detail
    errorStr += `Error is "${errorMessage}".\n`
  } else if ((is5xxError(error) || isTimeout(error)) && !failIfUnavailable) {
    errorStr += "Use the '--fail-if-unavailable' option to fail the command in this situation.\n"
  }

  return chalk.red(errorStr)
}

export const renderEvaluationRetry = (attempt: number, error: any): string => {
  if (is5xxError(error)) {
    const errorStatus = getStatus(error)

    return chalk.yellow(`[attempt ${attempt}] Gate evaluation failed with status code ${errorStatus}, retrying.\n`)
  }

  return chalk.yellow(`[attempt ${attempt}] Gate evaluation failed, retrying.\n`)
}

export const renderWaiting = (): string => {
  return 'Waiting for events to arrive...\n'
}
