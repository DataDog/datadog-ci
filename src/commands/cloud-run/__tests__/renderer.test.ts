import process from 'process'

import {MOCK_CWD} from '../../../helpers/__tests__/fixtures'
import {renderAdditionalFiles, renderProjectFiles} from '../../../helpers/renderer'

import {renderAuthenticationInstructions} from '../renderer'

process.cwd = jest.fn().mockReturnValue(MOCK_CWD)

describe('renderer', () => {
  test('renderAuthenticationInstructions', () => {
    expect(renderAuthenticationInstructions()).toMatchSnapshot()
  })

  describe('renderProjectFiles', () => {
    it('returns correct text when no project files were found', () => {
      expect(renderProjectFiles(new Set())).toMatchSnapshot()
    })

    it('returns correct text when multiple project files are found', () => {
      const mockFiles = new Set(['package.json', 'yarn.lock', 'tsconfig.json'])
      expect(renderProjectFiles(mockFiles)).toMatchSnapshot()
    })
  })

  describe('renderAdditionalFiles', () => {
    it('returns correct text when no additional files were added', () => {
      expect(renderAdditionalFiles(new Set())).toMatchSnapshot()
    })

    it('returns correct text when an additional file is added', () => {
      const mockFiles = new Set(['README.md'])
      expect(renderAdditionalFiles(mockFiles)).toMatchSnapshot()
    })
  })
})
