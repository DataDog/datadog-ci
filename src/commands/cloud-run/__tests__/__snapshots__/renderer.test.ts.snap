// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer renderAuthenticationInstructions 1`] = `
"
[Error] Unable to authenticate with GCP.

To authenticate with GCP, please follow these steps:
1. If you haven't already, install the Google Cloud SDK: https://cloud.google.com/sdk/docs/install
2. Run \\"gcloud auth application-default login\\" and follow the prompts in your browser to log in.
3. After logging in, run the \`datadog-ci cloud-run flare\` command again.
"
`;

exports[`renderer renderProjectFiles returns correct text when multiple project files are found 1`] = `
"
✅ Found project file(s) in mock-folder:
• package.json
• yarn.lock
• tsconfig.json
"
`;

exports[`renderer renderProjectFiles returns correct text when no project files were found 1`] = `
"[!] No project files found.
"
`;
